### ISSUE TITLE
Proposal: Typed CSS properties

### ISSUE THREAD

**Issue by MartinKavik** [seed-rs/seed#211]
Discussion in https://github.com/David-OConnor/seed/pull/209#discussion_r319706950:

> David-OConnor : 
Related: What do y'all think of having an `St` enum: An equivalent to `At`, but for styles? The groundwork's been there in a WIP state for a while; ended up not continuing due to the [large number of possible styles](https://developer.mozilla.org/en-US/docs/Web/CSS/Reference). I don't think it would hurt anything, but haven't heard a suggestion about this. The change would be to fill out [the make_styles!](https://github.com/David-OConnor/seed/blob/master/src/dom_types.rs#L340) macro call, add `St` to the prelude, and document it.

> tiberiusferreira: 
I really like the idea, but I wonder if the increase in binary size will be negligible, after all, the text will be stored (uncompressed before gzip, I imagine) somewhere in the binary. Anyway, I'd rather have the convenience even if at the cost of some KBs.

**Comment by MartinKavik** [comment]
> I wonder if the increase in binary size will be negligible

I was experimenting a little bit with [MartinKavik/seed-quickstart-webpack](https://github.com/MartinKavik/seed-quickstart-webpack) - there is a generator (PostCSS plugin) that creates Rust code from CSS classes. And because there is also TailwindCSS, the generated file `.rs` is big (86k lines).
However it seems that it doesn't affect the size of the `.wasm` file - it doesn't matter if that generated file has a struct with 5000 fields or 1, built `.wasm` always has 182 kb (~ 80kb compressed). 

tl;dr:
So I think that Rust is able to optimize it and adds only used values into the output. But we should test it once we add typed CSS into Seed for sure.

P.S.
Interesting fact - when I add `log!` into code, `wasm` file size is increased by ~200kb. We can investigate it / write into docs in the future.

**Comment by MartinKavik** [comment]
> The groundwork's been there in a WIP state for a while; ended up not continuing due to the large number of possible styles.

Yeah, it's definitely work for a generator.
Source list can be https://www.w3.org/Style/CSS/all-properties.en.html (there are more formats, see the end of the page) or https://developer.mozilla.org/en-US/docs/MDN/Contribute/Howto/Update_the_CSS_JSON_DB.

I suggest to find similar source lists for other typed entities - attributes, events and tags.
So we can write generators for all at once and probably refactor it. Ideally we would run something like `cargo make update_types` to regenerate types for all entities.

---

My opinion on typed CSS props - I would like to have even typed values..  (I don't know if it's possible and would be usable in practice). So I think it will be a good step forward.

**Comment by David-OConnor** [comment]
Agree on typed values; would make it so the IDE/compiler could tell if styles are valid in cases where only specific ones are allowed.

**Comment by David-OConnor** [comment]
Working on a parser for the JSON endpoint of the W3 site you linked. (Might switch to MDN). I think we can have it generate a rust file that's an expanded version of the `make_styles` call.

**Comment by MartinKavik** [comment]
> Agree on typed values;

It won't be easy to design API for it, collect source data and implement it. Let's focus on it somewhere in the future. (But I'm glad you agree.)

> Working on a parser for the JSON endpoint

Could you describe your plan/design a little bit (i.e. how will it work, libraries, API, etc.) before you dive into code?
Should I try to find a source data for attributes/events/tags so we can make it universal?

**Comment by David-OConnor** [comment]
- Add `Reqwest` as a dev dependency
- Pull JSON data from that API, and stores the properties in a text file that's gitignored

I'm not sure how to directly convert the parsed text file to a direct macro call like `make_styles` - any ideas?

Agree on leaving values for later.

If you find an API to pull attributes/events/tags, we could do the same. They're easier to deal with by hand since there's few of them, and don't update frequently, but would be a good addition.

**Comment by David-OConnor** [comment]
On second thought, I think the solution is to generate a rust file which we important the styles (etc) from.

**Comment by MartinKavik** [comment]
> If you find an API to pull attributes/events/tags, we could do the same

I think about creating own endpoint(s), because it's a mess. It shouldn't be hard (see https://zeit.co/blog/introducing-now-rust) and it can be a hard-coded JSON for the beginning. So the scraping / transforming / caching and similar ugly things would be in cloud. And we can define endpoint API and add some data for values later. I think it will be also useful for more people - we can create landing / about page with example of usage with Seed in the future.

> think the solution is to generate a rust file which we important the styles (etc) from.

Yeah, we can try to write a simple task with Rust script like https://github.com/David-OConnor/seed/blob/master/Makefile.toml#L245. If it turns out to be too complicated, we can write simple console application in a standalone create or something similar.
We would run this task before new Seed release.

What do you think?

**Comment by David-OConnor** [comment]
Down to make and host a JSON API.

I'm not great with Cargo Make. I can think of how to do with with a standalone bin crate or Python script, but ideally we'd run `cargo make make_styles` etc. I'll make a draft PR - want to take a crack at the makefile part after?

Agree on doing this before release; adding it to the release checklist.

**Comment by MartinKavik** [comment]
> want to take a crack at the makefile part after?

Yeah, I'll take it. 

If we want to use `https://zeit.co/` (I don't have experience with it, but it seems good for our requirements), I suggest to do it in those steps:

1. Write example files which we want to generate (probably something like `generated/styles.rs`, `generated/events.rs`, ..) and integrate them into Seed (i.e. extract current macros into standalone files and refactor it).

2. Write tasks for generation (against mocked endpoints).

3. Save JSON data from mocked endpoints into `json` files and deploy them to `zeit`.

4. Redirect tasks to the live endpoints (`json` files).

5. Write lambda function which scrape data about css props/tags/events/attributes and save them into hosted `json` files.

As the result, we will be able to download that JSON files how many times we want (and other people can use them too) and if something fails during scraping, it doesn't affect main endpoints (`json` files). 

So I'll take 2) and you probably 1). Pick other ones if/which you want, I'll take remainder.  

What do you think?

**Comment by David-OConnor** [comment]
Draft PR linked above. If unable to get working with makefile, can refactor into a separate crate (Maybe workspace?) or script.

**Comment by MartinKavik** [comment]
Looks good at the first glance, I'll try to integrate it into a new makefile task at the evening / night (it's 13.00 in CZ now) and I'll send PR to your `style_numes` branch.

**Comment by tiberiusferreira** [comment]
> P.S.
> Interesting fact - when I add `log!` into code, `wasm` file size is increased by ~200kb. We can investigate it / write into docs in the future.

This is quite a lot of added weight. I think it is reasonable to expect apps to regularly use log! facilities. Is there a way to disable the log! usage when deploying the app? (without manually commenting out the code).

**Comment by MartinKavik** [comment]
@tiberiusferreira I forgot where I added that `log!`, but I cannot reproduce it now. So it's maybe resolved or not critical.

**Comment by MartinKavik** [comment]
`St` is typed and its variants are generated from our endpoint. So I think that it resolves this issue.
New style API will be designed here: https://github.com/seed-rs/seed/issues/332
